# 4-Manifolds and Log VOA
## Example section companion notebook

Seifert manifolds with four singular fibers can be related to characters of $\mathcal V(p,p')$ VOAs. Theorem 4.7 provides a detailed description of this relation for spherical Seifert manifolds. In this example notebook, we replicate the computations done in writing the first paragraph of section 6.2 of arXiv:2201.04640, while at the same time demonstrating the capabilities and main functions of pySeifert: a Sage toolkit for the computation of $\hat{Z}$ invariants for Seifert manifolds and related q-series invariants.

We start by loading pySeifert and defining a Seifert manifold

In [1]:
load("../../pySeifert.sage")
S = Seifert([-2, 1, 2, 2, 3, 2, 5, 3, 7]);S

Seifert manifold with 4 exceptional fibers.
Seifert data:
[-2, 1, 2, 2, 3, 2, 5, 3, 7]
Plumbing Matrix:
[-2  1  1  0  1  0  1  0  0]
[ 1 -2  0  0  0  0  0  0  0]
[ 1  0 -2  1  0  0  0  0  0]
[ 0  0  1 -2  0  0  0  0  0]
[ 1  0  0  0 -3  1  0  0  0]
[ 0  0  0  0  1 -2  0  0  0]
[ 1  0  0  0  0  0 -3  1  0]
[ 0  0  0  0  0  0  1 -2  1]
[ 0  0  0  0  0  0  0  1 -2]
D: 1, m: 210, det(M): -1

The plumbing matrix is unitary therefore $S$ is a spherical manifold and the lattice dilation factor $D$ is one.
Becuase $S$ is spherical we expect to see only trivial boundary conditions.

In [2]:
B = S.boundary_conditions(["A",1],basis="root"); B

[[(-1), (1/2), (0), (1/2), (0), (1/2), (0), (0), (1/2)]]

Pairs $p$ and $p'$ should be chosen such that $pp'=m$ and $p=p_i$ from the Seifert data. In this case $m=210$, so the possible pairs of $p$, $p'$ are,:

In [4]:
p_pp_l = list()
for p in S.p:
    p_pp_l.append([p,S.m/p])
p_pp_l

[[2, 105], [3, 70], [5, 42], [7, 30]]

To compare the $\hat Z$ integrand to the triplet characters it is useful to list all possible $\hat w$ along with their lengths.

In [6]:
WG = weyl_group(["A",1])
WL = [det(g) for g in WG]
wh_l = list(itertools.product(*[WG if d == 1 else [matrix(1)] for d in S.deg]))
wl_l = list(itertools.product(*[WL if d == 1 else [1] for d in S.deg]))

For each choice of $p,p'$ we can compute the linear combination of $\log-\mathcal{V}(p,p')$ characters (as in equation 4.54) and compare to the $\hat{Z}$ invariant

In [13]:
n = 0
p = S.p[n]
pp = S.m/p
r = 1
Chp = 0
Trip = 0
expMax = 100
delta = S.delta(["A",1])
s_list = list()
for wh,l in zip(wh_l,wl_l):
    l_red = [li for li,d in zip(l,S.deg) if d == 1]
    Chp += product(l_red)*S.chi_prime_4f_sph(wh,expMax,basis="root")
    if l_red[n] == 1:
        s = abs(pp*(sum([li*sgn(qi)/pi for pi,qi,li in zip(S.p,S.q,l_red) if pi != p])))
        epss = sign(S.q[n]*sum([li*sgn(qi)/pi for pi,qi,li in zip(S.p,S.q,l_red) if pi != p]))
        s_list.append(abs(s))
        Trip += product(l_red)*epss*triplet_character_p_pprime(p,pp,s,r,expMax-delta)
print(f"p: {p}")
print(f"p': {pp}")
print(f"Ẑ integrand:\n {Chp}")
print(f"Triplet character:\n {Trip}")
print(f"Ẑ integrand - q^delta * triplet character: {expand(Chp-q^delta*Trip)}")

p: 2
p': 105
Ẑ integrand:
 -2*q^(317/4)*z^2/(z^2 + 1/z^2 - 2) + 4*q^(317/4)/(z^2 + 1/z^2 - 2) - 2*q^(317/4)/((z^2 + 1/z^2 - 2)*z^2) - 2*q^(261/4)*z^2/(z^2 + 1/z^2 - 2) + 4*q^(261/4)/(z^2 + 1/z^2 - 2) - 2*q^(261/4)/((z^2 + 1/z^2 - 2)*z^2) + 2*q^(145/4)*z^2/(z^2 + 1/z^2 - 2) - 4*q^(145/4)/(z^2 + 1/z^2 - 2) + 2*q^(145/4)/((z^2 + 1/z^2 - 2)*z^2)
Triplet character:
 -2*(z^2 + 1/z^2 - 2)*q^(66049/840)/(z - 1/z)^2 - 2*(z^2 + 1/z^2 - 2)*q^(54289/840)/(z - 1/z)^2 + 2*(z^2 + 1/z^2 - 2)*q^(29929/840)/(z - 1/z)^2
Ẑ integrand - q^delta * triplet character: 0


In [12]:
n = 1
p = S.p[n]
pp = S.m/p
r = 1
Chp = 0
Trip = 0
expMax = 100
delta = S.delta(["A",1])
s_list = list()
for wh,l in zip(wh_l,wl_l):
    l_red = [li for li,d in zip(l,S.deg) if d == 1]
    Chp += product(l_red)*S.chi_prime_4f_sph(wh,expMax,basis="root")
    if l_red[n] == 1:
        s = abs(pp*(sum([li*sgn(qi)/pi for pi,qi,li in zip(S.p,S.q,l_red) if pi != p])))
        epss = sign(S.q[n]*sum([li*sgn(qi)/pi for pi,qi,li in zip(S.p,S.q,l_red) if pi != p]))
        s_list.append(abs(s))
        Trip += product(l_red)*epss*triplet_character_p_pprime(p,pp,s,r,expMax-delta)
print(f"p: {p}")
print(f"p': {pp}")
print(f"Ẑ integrand:\n {Chp}")
print(f"Triplet character:\n {Trip}")
print(f"Ẑ integrand - q^delta * triplet character: {expand(Chp-q^delta*Trip)}")

p: 3
p': 70
Ẑ integrand:
 -2*q^(317/4)*z^2/(z^2 + 1/z^2 - 2) + 4*q^(317/4)/(z^2 + 1/z^2 - 2) - 2*q^(317/4)/((z^2 + 1/z^2 - 2)*z^2) - 2*q^(261/4)*z^2/(z^2 + 1/z^2 - 2) + 4*q^(261/4)/(z^2 + 1/z^2 - 2) - 2*q^(261/4)/((z^2 + 1/z^2 - 2)*z^2) + 2*q^(145/4)*z^2/(z^2 + 1/z^2 - 2) - 4*q^(145/4)/(z^2 + 1/z^2 - 2) + 2*q^(145/4)/((z^2 + 1/z^2 - 2)*z^2)
Triplet character:
 -2*(z^2 + 1/z^2 - 2)*q^(66049/840)/(z - 1/z)^2 - 2*(z^2 + 1/z^2 - 2)*q^(54289/840)/(z - 1/z)^2 + 2*(z^2 + 1/z^2 - 2)*q^(29929/840)/(z - 1/z)^2
Ẑ integrand - q^delta * triplet character: 0


In [14]:
n = 2
p = S.p[n]
pp = S.m/p
r = 1
Chp = 0
Trip = 0
expMax = 100
delta = S.delta(["A",1])
s_list = list()
for wh,l in zip(wh_l,wl_l):
    l_red = [li for li,d in zip(l,S.deg) if d == 1]
    Chp += product(l_red)*S.chi_prime_4f_sph(wh,expMax,basis="root")
    if l_red[n] == 1:
        s = abs(pp*(sum([li*sgn(qi)/pi for pi,qi,li in zip(S.p,S.q,l_red) if pi != p])))
        epss = sign(S.q[n]*sum([li*sgn(qi)/pi for pi,qi,li in zip(S.p,S.q,l_red) if pi != p]))
        s_list.append(abs(s))
        Trip += product(l_red)*epss*triplet_character_p_pprime(p,pp,s,r,expMax-delta)
print(f"p: {p}")
print(f"p': {pp}")
print(f"Ẑ integrand:\n {Chp}")
print(f"Triplet character:\n {Trip}")
print(f"Ẑ integrand - q^delta * triplet character: {expand(Chp-q^delta*Trip)}")

p: 5
p': 42
Ẑ integrand:
 -2*q^(317/4)*z^2/(z^2 + 1/z^2 - 2) + 4*q^(317/4)/(z^2 + 1/z^2 - 2) - 2*q^(317/4)/((z^2 + 1/z^2 - 2)*z^2) - 2*q^(261/4)*z^2/(z^2 + 1/z^2 - 2) + 4*q^(261/4)/(z^2 + 1/z^2 - 2) - 2*q^(261/4)/((z^2 + 1/z^2 - 2)*z^2) + 2*q^(145/4)*z^2/(z^2 + 1/z^2 - 2) - 4*q^(145/4)/(z^2 + 1/z^2 - 2) + 2*q^(145/4)/((z^2 + 1/z^2 - 2)*z^2)
Triplet character:
 -2*(z^2 + 1/z^2 - 2)*q^(66049/840)/(z - 1/z)^2 - 2*(z^2 + 1/z^2 - 2)*q^(54289/840)/(z - 1/z)^2 + 2*(z^2 + 1/z^2 - 2)*q^(29929/840)/(z - 1/z)^2
Ẑ integrand - q^delta * triplet character: 0


In [15]:
n = 3
p = S.p[n]
pp = S.m/p
r = 1
Chp = 0
Trip = 0
expMax = 100
delta = S.delta(["A",1])
s_list = list()
for wh,l in zip(wh_l,wl_l):
    l_red = [li for li,d in zip(l,S.deg) if d == 1]
    Chp += product(l_red)*S.chi_prime_4f_sph(wh,expMax,basis="root")
    if l_red[n] == 1:
        s = abs(pp*(sum([li*sgn(qi)/pi for pi,qi,li in zip(S.p,S.q,l_red) if pi != p])))
        epss = sign(S.q[n]*sum([li*sgn(qi)/pi for pi,qi,li in zip(S.p,S.q,l_red) if pi != p]))
        s_list.append(abs(s))
        Trip += product(l_red)*epss*triplet_character_p_pprime(p,pp,s,r,expMax-delta)
print(f"p: {p}")
print(f"p': {pp}")
print(f"Ẑ integrand:\n {Chp}")
print(f"Triplet character:\n {Trip}")
print(f"Ẑ integrand - q^delta * triplet character: {expand(Chp-q^delta*Trip)}")

p: 7
p': 30
Ẑ integrand:
 -2*q^(317/4)*z^2/(z^2 + 1/z^2 - 2) + 4*q^(317/4)/(z^2 + 1/z^2 - 2) - 2*q^(317/4)/((z^2 + 1/z^2 - 2)*z^2) - 2*q^(261/4)*z^2/(z^2 + 1/z^2 - 2) + 4*q^(261/4)/(z^2 + 1/z^2 - 2) - 2*q^(261/4)/((z^2 + 1/z^2 - 2)*z^2) + 2*q^(145/4)*z^2/(z^2 + 1/z^2 - 2) - 4*q^(145/4)/(z^2 + 1/z^2 - 2) + 2*q^(145/4)/((z^2 + 1/z^2 - 2)*z^2)
Triplet character:
 -2*(z^2 + 1/z^2 - 2)*q^(66049/840)/(z - 1/z)^2 - 2*(z^2 + 1/z^2 - 2)*q^(54289/840)/(z - 1/z)^2 + 2*(z^2 + 1/z^2 - 2)*q^(29929/840)/(z - 1/z)^2
Ẑ integrand - q^delta * triplet character: 0
